File: 287-find-the-duplicate-number/287-find-the-duplicate-number.cpp
class Solution {
public:
    int findDuplicate(vector<int>& nums) {
        unordered_set<int> seen;
        for(auto x : nums){
            if(seen.count(x)) 
                return x;
            seen.insert(x);
        }
        return -1;
    }
};

File: 287-find-the-duplicate-number/287-find-the-duplicate-number.py
class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        
        fast = nums[0]
        slow = nums[0]
        
        while True: #since python doesn't have do-while
            slow = nums[slow]
            fast = nums[nums[fast]]
            if slow == fast:
                break
        
        fast = nums[0]
        
        while fast != slow:
            slow = nums[slow]
            fast = nums[fast]
        
        return fast